azurelinuxagent/ga/cgroupapi.py

```diff
@@ -342,11 +397,139 @@ def start_extension_command(self, extension_name, command, cmd_name, timeout, sh
         """
-        return CGroupsApi._foreach_legacy_cgroup(lambda *_: None)
+        if cgroup_path is not None and controller is not None:
+            # Check that the controller is enabled in the cgroup.subtree_control file
+            enabled_controllers_file = os.path.join(cgroup_path, 'cgroup.subtree_control')
+            if os.path.exists(enabled_controllers_file):
+                enabled_controllers = fileutil.read_file(enabled_controllers_file).rstrip().split(" ")
+                if controller in enabled_controllers:
+                    return True
+
+            # Check that the controller interface files exist in the cgroup
+            if os.path.exists(cgroup_path):
+                for item in os.listdir(cgroup_path):
+                    if item.startswith(controller + '.'):
+                        return True
+
+        return False
+
+    def get_cgroup_mount_points(self):
+        # The output of mount is similar to
+        #     $ findmnt -t cgroup2 --noheadings
+        #     /sys/fs/cgroup cgroup2 cgroup2 rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot
+        #
+        # Since v2 is a unified hierarchy, this method checks if each controller is enabled at the root cgroup. This
+        # check is necessary because all non-root "cgroup.subtree_control" files can only contain controllers which are
+        # enabled in the parent's "cgroup.subtree_control" file.
+
+        if not self._cgroup_mountpoints:
+            cpu = None
+            memory = None
+            for line in shellutil.run_command(['findmnt', '-t', 'cgroup2', '--noheadings']).splitlines():
+                match = re.search(r'(?P<path>/\S+)\s+cgroup2', line)
+                if match is not None:
+                    mount_point = match.group('path')
+                    if self.is_controller_enabled('cpu', mount_point):
+                        cpu = mount_point
+                    if self.is_controller_enabled('memory', mount_point):
+                        memory = mount_point
+            self._cgroup_mountpoints = {'cpu': cpu, 'memory': memory}
+
+        return self._cgroup_mountpoints['cpu'], self._cgroup_mountpoints['memory']
+
+    def get_unit_cgroup_paths(self, unit_name):
+        # Ex: ControlGroup=/azure.slice/walinuxagent.service
+        #     controlgroup_path[1:] = azure.slice/walinuxagent.service
+        controlgroup_path = systemd.get_unit_property(unit_name, "ControlGroup")
+        cpu_mount_point, memory_mount_point = self.get_cgroup_mount_points()
+
+        # Since v2 is a unified hierarchy, we need to check if each controller is enabled for the cgroup. If a
+        # controller is not enabled, then its controller interface files won't exist at the cgroup path
+        cpu_cgroup_path = None
+        if cpu_mount_point is not None:
+            cgroup_path = os.path.join(cpu_mount_point, controlgroup_path[1:])
+            if self.is_controller_enabled('cpu', cgroup_path):
```

Comment:
why do we need to check this here? This method should return what api returns?

Reply:
the v1 api gives us different paths for each controller in use by the cgroup. the v2 api result doesn't tell us if the controller is enabled for the cgroup, because it just returns one path at which any controller might be enabled. 

later we use this method to get the cgroup_paths for tracking. if cpu/memory isn't enabled, then we initialize CpuCgroup or MemoryCgroup with path that does have corresponding counter interface files.

Reply:
we call these paths for tracking and other purposes once we start usingthe cgroups for v2 that means we enable the controllers already. why do we need extra checks? or maybe I'm missing something here.

Reply:
we attempt to enable the controllers by adding drop-in files, updating slice, etc., but we don't have an explicit check that they were successfully enabled and that the controllers are in use

I think there should be an explicit check, if not here then somewhere else. I added the check here because in v1 we know the cgroup is in cpu/memory because path is provided. in v2 we're not guaranteed the same

Reply:
will remove this check and only check if controllers are enabled at root